### PR TITLE
refactor(vmstorage): Refactor the code to reduce the time complexity of `MustAddRows` and improve readability

### PR DESCRIPTION
**Pull request by tanruixiang**
### Describe Your Changes
The original logic is not only highly complex but also poorly readable, so it can be modified to increase readability and reduce time complexity.

### Checklist

The following checks are **mandatory**:

- [x] My change adheres [VictoriaMetrics contributing guidelines](https://docs.victoriametrics.com/contributing/).

### CODE REVIEW

**Comment by tanruixiang** [comment]
@jiekun Hi, I changed the git message and I see you tagged my pr before, can you rerun the CI and review this pr? Thank you very much.

**Comment by jiekun** [comment]
The code in this pull request appears to be more concise. The original one has a chance to break earlier and probably doesn't need to go through all the rows. Any thought on this? @f41gh7

File: lib/storage/table.go [hunk 1]
```diff
@@ -267,22 +267,20 @@ func (tb *table) MustAddRows(rows []rawRow) {
 	if len(rows) == 0 {
 		return
 	}
-
+	// Get the maximum and minimum timestamps in the rows
+	maxTsInRows, minTsInRows := rows[0].Timestamp, rows[0].Timestamp
+	for i := range rows {
```
Comment (jiekun) [inline lib/storage/table.go:272]: The code in this pull request appears to be more concise. The original one has a chance to break earlier and probably doesn't need to go through all the rows. Any thought on this? @f41gh7

Comment (f41gh7) [reply lib/storage/table.go:272]: Break earlier seems to be a rare case for cross partition data ingestion.

Currently, data partitioned by month. And it should be a rare case, when a single data block has timestamps for multiple months. 

 So for me it looks like valid optimization. Most common case is single partition data ingestion, it'll not trigger previous version fast path.

 Also It'd be great to have a bench for it.

**Comment by tanruixiang** [comment]
After discussing with @jiekun , he suggested me to test the case where all the timestamps of the written data belong to one partition. And to further improve the credibility of the optimization give specific elapsed time. Many thanks to @jiekun  for his help.
In this case, analyzing the time complexity, one is O(n) and the other is O(nm), and undoubtedly O(n) is faster.
I tested this with the code below and the results were very much in line with the time complexity, with the original method taking about 17us on average and the optimized method taking 150ns on average.

```
func MockData() ([]rawRow, []*partitionWrapper) {
	timestamp := 1721286843000
	pt := newPartition("test", "smallPartsPath", "bigPartsPath", nil)
	pt.tr.fromPartitionTimestamp(int64(timestamp))
	ptw := partitionWrapper{pt: pt}
	ptws := []*partitionWrapper{&ptw}
	mockRows := make([]rawRow, 0)
	for i := range 20000 {
		mockRows = append(mockRows, rawRow{Timestamp: 1721286843000 - int64(i)})
	}
	return mockRows, ptws
}
func TestMustAddRows() {
	mockRows, ptws := MockData()
	startTime := time.Now()
	for _, ptw := range ptws {
		singlePt := true
		for j := range mockRows {
			if !ptw.pt.HasTimestamp(mockRows[j].Timestamp) {
				singlePt = false
				break
			}
		}
		if !singlePt {
			panic("test error")
		}
	}
	elapsedTime := time.Since(startTime)
	fmt.Printf("original time: %+v\n", elapsedTime)
}

func TestMustAddRowsTimeRange() {
	mockRows, ptws := MockData()
	// Get the maximum and minimum timestamps in the rows
	maxTsInRows, minTsInRows := mockRows[0].Timestamp, mockRows[0].Timestamp
	for i := range mockRows {
		maxTsInRows = max(maxTsInRows, mockRows[i].Timestamp)
		minTsInRows = min(minTsInRows, mockRows[i].Timestamp)
	}
	startTime := time.Now()
	for _, ptw := range ptws {
		if !ptw.pt.TimeRangeInPartition(minTsInRows, maxTsInRows) {
			panic("test error")
		}
	}
	elapsedTime := time.Since(startTime)
	fmt.Printf("use time range time: %+v\n", elapsedTime)
}
```

**Review by f41gh7 — APPROVED** [review]
LGTM

Please update changelog

**Comment by tanruixiang** [comment]
> LGTM
> 
> Please update changelog

@f41gh7 Thank you so much for your help.
I noticed that the branch for the cluster version is on `cluster`, do I need to merge the changes into the cluster version at another pr?

**Comment by jiekun** [comment]
@tanruixiang No. It's maintainers responsibility to handle the merge from feature branch to cluster/enterprise/lts branches.

**Comment by tanruixiang** [comment]
> @tanruixiang No. It's maintainers responsibility to handle the merge from feature branch to cluster/enterprise/lts branches.

Got it, thanks for the reply.

**Comment by jiekun** [comment]
> So my changlog only needs to describe the single-node version

No, the change log should include description for **all affected components**. However, the maintainer will submit follow-up commits to correct the necessary information, if needed.

**Comment by tanruixiang** [comment]
> No, the change log should include description for **all affected components**. However, the maintainer will submit follow-up commits to correct the necessary information, if needed.

Thank you very much!
 I've added a description of the cluster version to the changelog.

File: docs/CHANGELOG.md [hunk 2]
```diff
@@ -35,6 +35,7 @@ See also [LTS releases](https://docs.victoriametrics.com/lts-releases/).
 
 * SECURITY: upgrade base docker image (Alpine) from 3.20.1 to 3.20.2. See [alpine 3.20.2 release notes](https://alpinelinux.org/posts/Alpine-3.20.2-released.html).
 
+* FEATURE: `vmstorage` in the [Single-node VictoriaMetrics](https://docs.victoriametrics.com/) and the [VictoriaMetrics cluster](https://docs.victoriametrics.com/cluster-victoriametrics/): Refactor the code located in the `MustAddRows` function that looks for the corresponding partition when data is inserted. The new code efficiently determines whether all data to be inserted belongs to the corresponding partition by obtaining the maximum and minimum timestamps of all data to be inserted, resulting in better performance and readability.
```
Comment (jiekun) [inline docs/CHANGELOG.md:38]: Those code changes look good to me.

```suggestion
* FEATURE: [VictoriaMetrics Single-Node](https://docs.victoriametrics.com/) and [VictoriaMetrics Cluster](https://docs.victoriametrics.com/cluster-victoriametrics/): Refactor the code located in the `MustAddRows` function that looks for the corresponding partition when data is inserted. The new code efficiently determines whether all data to be inserted belongs to the corresponding partition by obtaining the maximum and minimum timestamps of all data to be inserted, resulting in better performance and readability.
```

per some discussion in internal channel:
- VictoriaMetrics Single-Node (or Single Node)
- VictoriaMetrics Cluster

**Review by jiekun — APPROVED** [review]

**Comment by tanruixiang** [comment]
@jiekun Thank you. I've merged your suggestions. I noticed that `MustAddRows` exist in more than just `vmstorage`, so I further clarified in the changlog that `MustAddRows` are located in `vmstorage`.

**Comment by f41gh7** [comment]
> > LGTM
> > Please update changelog
> 
> @f41gh7 Thank you so much for your help. I noticed that the branch for the cluster version is on `cluster`, do I need to merge the changes into the cluster version at another pr?

There is no need for that. Any changes to master branch back-ported to the other branches by VictoriaMetrics team later.

**Comment by f41gh7** [comment]
Thanks for contribution!

**Comment by tanruixiang** [comment]
> Thanks for contribution!

@f41gh7 @jiekun Thank you for your help!

**Comment by valyala** [comment]
@tanruixiang , @f41gh7 , @jiekun , this pull request has been reverted in the commit 8551fbe9f38872fee829a9a821911ae9307ec2e7 . Please read the commit description for the reasons of the revert.

**Pull request merged**